### PR TITLE
Add Agent to support RTSP compatible IP cameras

### PR DESCRIPTION
**Pull request by tskisner**
This agent adds support for acquiring snapshots and short recordings from cameras that can stream the RTSP protocol.

## Description

This new agent records frames to disk (default every 10s) and passes them through a motion detection algorithm.  If there is motion, a 60 second recording is taken.  Recordings can also be triggered directly as an agent Task.  Both images and video are stored to disk in circular buffers which by default hold the most recent 2 days of snapshots and 120 minutes of triggered videos.  Use of this agent requires the opencv / cv2 and imutils packages.

Note that each agent instance handles exactly one camera.  This enables using different settings for different camera placements.

## Motivation and Context

This agent was created to support the Dahua security cameras at the site.  These are interior cameras designed to capture unauthorized access or theft.

## How Has This Been Tested?

This agent has been tested on a local system (laptop) at the site.  It has been run for several hours and successfully detected motion in one of the containers as people went in and out.  Note that I only tested this running as a local agent on the host system.  I verified that the CPU and memory use are not too dramatic.  There is a brief spike at the end of a recording when the list of frames is encoded to mp4.  Here is an example of motion detection, which draws a green box around the areas where a change was detected:

img_2023-12-30T16:54:53+00:00
img_2023-12-30T16:55:03+00:00

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by tskisner** [comment]
Just a note:  this agent does require the `opencv` and `imutils` packages.  Let me know if you want me to add another "extra" key to setup.py for this case.

File: socs/agents/camera_rtsp/agent.py
Comment (davidvng) [inline socs/agents/camera_rtsp/agent.py:565]: Is there any way to increase the recording duration if motion is continuously being detected? I figure we'd want to have a continuous video of a break-in that occurs longer than 60 seconds rather than having to open many 60 second videos.

File: socs/agents/camera_rtsp/agent.py
Comment (davidvng) [inline socs/agents/camera_rtsp/agent.py:435]: Can you make a similar argument that disables motion based on some time range during the day? I figure we wouldn't need to record motion during work hours when people on site are continuously entering/exiting containers

File: socs/agents/camera_rtsp/agent.py
Comment (davidvng) [inline socs/agents/camera_rtsp/agent.py:94]: I'm not too familiar with circular media buffers, pardon my ignorance here: what happens when the agent crashes/exits and is restarted? Does this gracefully pick back up the buffer and continue handling files correctly?

File: socs/agents/camera_rtsp/agent.py
Comment (davidvng) [inline socs/agents/camera_rtsp/agent.py:387]: Nitpicky, but could you update this and all other instances to `RTSPCameraAgent` for consistency with the `ACTiCameraAgent`?

**Review by davidvng — CHANGES_REQUESTED** [review]
Thanks so much for writing this agent! Here's my initial review with some requests. Mainly name changes and core functionality requests.

I'm also wondering if the classes `CircularMediaBuffer` and `MotionDetector` should become external classes since they can also be used by the `ACTiCameraAgent` and any other similar agents in the future. Though this can be a discussion with others.

**Comment by BrianJKoopman** [comment]
This is great. Thanks for reviewing @davidvng, and thanks for the PR @tskisner! 

> Just a note: this agent does require the `opencv` and `imutils` packages. Let me know if you want me to add another "extra" key to setup.py for this case.

This would be good to do. If these will be used in the functions that @davidvng suggested get pulled out for use in the ACTi camera agent then I don't mind them going in the main `install_requires`, else a new `extras_require` group would be great.

Since I assume you're busy with other things now that you're back from the site @tskisner, let's go ahead and merge this as is, and @davidvng will take over making these improvements in separate PRs. That'll allow us to stand-up the agent for the cameras already deployed on site, and move forward with these suggestions. Sound good?

File: socs/agents/camera_rtsp/agent.py
Comment (tskisner) [reply socs/agents/camera_rtsp/agent.py:762]: That sounds reasonable, I'll make it more descriptive.

File: socs/agents/camera_rtsp/agent.py
Comment (tskisner) [reply socs/agents/camera_rtsp/agent.py:565]: I'll see how difficult it is to change the logic.  In the worst case, we could increase the recording time to something longer and ensure that it is only enabled at night.

File: socs/agents/camera_rtsp/agent.py
Comment (tskisner) [reply socs/agents/camera_rtsp/agent.py:435]: That would be a nice feature that I considered for a future update, however I guess the usefulness of this agent is somewhat limited if we are continuously triggering off motion during the day.  I'll work on adding a time of day start / stop setting.

File: socs/agents/camera_rtsp/agent.py
Comment (tskisner) [reply socs/agents/camera_rtsp/agent.py:94]: Yes, on construction it scans the existing files and picks up where it left off.  I'll move these helper classes into `socs/common/camera.py` for general use.

File: socs/agents/camera_rtsp/agent.py
Comment (tskisner) [reply socs/agents/camera_rtsp/agent.py:387]: Sure, I was aiming for `CategorySpecificThingAgent`, but looking across the codebase now I see most agents are `SpecificThingCategoryAgent`.  I'll change it.

Comment (tskisner) [reply socs/agents/camera_rtsp/agent.py]: This is implemented.

File: socs/agents/camera_rtsp/agent.py
Comment (tskisner) [reply socs/agents/camera_rtsp/agent.py]: Done

File: socs/agents/camera_rtsp/agent.py
Comment (tskisner) [reply socs/agents/camera_rtsp/agent.py]: This is now implemented as `--motion_start` and `--motion_stop` options.  The format is an ISO style string with UTC offset.  For example, to only perform motion detection between 7pm and 6am Chile time, one would pass:
```
--motion_start 19:00:00-04:00
--motion_stop 06:00:00-04:00
```

File: socs/agents/camera_rtsp/agent.py
Comment (tskisner) [reply socs/agents/camera_rtsp/agent.py]: This is quite challenging to implement with the current logic and I am not going to do this now.  Hopefully it is sufficient to increase the recording time to a few minutes and only enable motion detection during the night.

**Comment by tskisner** [comment]
This is now ready for re-review @davidvng.

File: socs/agents/rtsp_camera/agent.py
Comment (davidvng) [inline socs/agents/rtsp_camera/agent.py:292]: Can you add `movement` bool to this? We may want to monitor on Grafana and trigger alarms if the cameras see movement.

File: socs/agents/rtsp_camera/agent.py
Comment (davidvng) [inline socs/agents/rtsp_camera/agent.py:236]: Does this `grab` have to be within the loop or can it just grab at the end? From my understanding, this is grabbing 200 images for example but only retrieving the last one.

File: socs/common/camera.py
Comment (davidvng) [inline socs/common/camera.py:245]: Can you add these arguments as optional kwargs to the agent? We may want to change these if the camera FOV, room size, etc cause motion detection to fail.

**Review by davidvng — COMMENTED** [review]
@tskisner Thanks for the updates! I added some comments below. Please also add this agent to [`plugin.py`](https://github.com/tskisner/socs/blob/rtsp_camera/socs/plugin.py).

File: tests/agents/test_camera_rtsp.py
Comment (BrianJKoopman) [reply tests/agents/test_camera_rtsp.py:1]: This sounds good, but I don't think we should let it hold up this PR at this point if it's not straightforward.

Comment (tskisner) [reply tests/agents/test_camera_rtsp.py:1]: Agreed, it is more important in my opinion to get this merged, so that at least there is a starting point for future development / extension.  I just rebased this branch and dealt with the setup.py to pyproject.toml transition.  Another example of why we should merge this if it does no harm.

File: socs/agents/rtsp_camera/agent.py
Comment (tskisner) [reply socs/agents/rtsp_camera/agent.py:236]: This is a consequence of the way these cameras work.  You have to continuously grab images and extract the ones you want.

**Review by BrianJKoopman — APPROVED** [review]
Made some small packaging/docs changes. Will squash and merge when checks complete.

**Pull request merged**
